python/ray/serve/_private/router.py

```diff
@@ -209,40 +232,65 @@ def target_num_scheduling_tasks(self) -> int:
 
         This will never exceed `self.max_num_scheduling_tasks`.
         """
-        return min(self.num_pending_assignments, self.max_num_scheduling_tasks)
+        return min(self.num_pending_requests, self.max_num_scheduling_tasks)
+
+    @property
+    def curr_replicas(self) -> Dict[str, ReplicaWrapper]:
+        return self._replicas
 
     def update_replicas(self, replicas: List[ReplicaWrapper]):
         """Update the set of available replicas to be considered for scheduling.
 
         When the set of replicas changes, we may spawn additional scheduling tasks
         if there are pending assignments.
         """
-        self._replicas = {r.replica_id: r for r in replicas}
-        new_replica_id_set = set(self._replicas.keys())
+        new_replicas = {}
+        new_replica_id_set = set()
+        new_multiplexed_model_id_to_replica_ids = defaultdict(set)
+        for r in replicas:
+            new_replicas[r.replica_id] = r
+            new_replica_id_set.add(r.replica_id)
+            for model_id in r.multiplexed_model_ids:
+                new_multiplexed_model_id_to_replica_ids[model_id].add(r.replica_id)
+
+        self._replicas = new_replicas
+        self._replica_id_set = new_replica_id_set
+        self._multiplexed_model_id_to_replica_ids = (
+            new_multiplexed_model_id_to_replica_ids
+        )
+        self._replicas_updated_event.set()
+        self.maybe_start_scheduling_tasks()
+
         if self._replica_id_set != new_replica_id_set:
-            self._replica_id_set = new_replica_id_set
             logger.info(
                 "Got updated replicas for deployment "
                 f"{self._deployment_name}: {new_replica_id_set}.",
                 extra={"log_to_stderr": False},
             )
 
-        self._replicas_updated_event.set()
-        self.maybe_start_scheduling_tasks()
-
     def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
         """Shim for compatibility with the existing round robin scheduler."""
         return self.update_replicas([ActorReplicaWrapper(r) for r in running_replicas])
 
     async def choose_two_replicas_with_backoff(
         self,
+        request_metadata: Optional[RequestMetadata] = None,
     ) -> AsyncGenerator[List[RunningReplicaInfo], None]:
         """Generator that repeatedly chooses two random replicas from `self._replicas`.
 
         After each iteration, there will be an increasing backoff sleep time (dictated
         by `self.backoff_sequence_s`). The caller should exit the generator to reset the
         backoff sleep time.
         """
+        # TODO: comment.
+        replica_ids_with_multiplexed_model_id = set()
+        if request_metadata is not None and request_metadata.multiplexed_model_id:
+            replica_ids_with_multiplexed_model_id = copy.copy(
```

Comment:
how do we guarantee that `self._multiplexed_model_id_to_replica_ids` must have the key `request_metadata.multiplexed_model_id`

Reply:
oh... we don't! let me fix...

Reply:
it currently works because it's a `defaultdict`, but i'll make an explicit `get` call instead